CMake ZLIB detection does not work  on some machines

For some reason `linuxlike/13 cmake dependency` fails to work for me on Kubuntu 18/04 and 18/10. It does work on a different machine with Ubuntu 18/04. I have CMake and stuff installed and `FindZLIB.cmake` is in the `/usr/share`.

Attached is a log of the output of CMake trace as it is run.
[cmakelookup.txt](https://github.com/mesonbuild/meson/files/2780319/cmakelookup.txt)

```
cmake --find-package -DCOMPILER_ID=GNU -DLANGUAGE=CXX -DMODE=EXIST -DNAME=ZLIB 
ZLIB not found.
CMake Error: Problem processing arguments. Aborting.
```

In this case, this is (almost certainly) a CMake issue:

Output for __found__ dependency:

```bash
~ cmake --find-package -DCOMPILER_ID=GNU -DLANGUAGE=CXX -DMODE=EXIST -DNAME=ZLIB
ZLIB found.
```

Output for __not found__ dependency:

```bash
~ cmake --find-package -DCOMPILER_ID=GNU -DLANGUAGE=CXX -DMODE=EXIST -DNAME=DoesNotExist
DoesNotExist not found.
CMake Error: Problem processing arguments. Aborting.

```

--- a/mesonbuild/dependencies/data/CMakeLists.txt
+++ b/mesonbuild/dependencies/data/CMakeLists.txt
@@ -18,8 +18,7 @@ string(TOUPPER "${_packageName}" PACKAGE_NAME)
 while(TRUE)
   find_package("${NAME}" QUIET)
 
-  list(LENGTH LIB_ARCH_LIST LIST_LEN)
-  if(${_packageName}_FOUND OR ${PACKAGE_NAME}_FOUND OR LIST_LEN LESS_EQUAL 0)
+  if(${_packageName}_FOUND OR ${PACKAGE_NAME}_FOUND OR "${LIB_ARCH_LIST}" STREQUAL "")
     break()
   endif()
 

--- a/mesonbuild/dependencies/data/CMakeLists.txt
+++ b/mesonbuild/dependencies/data/CMakeLists.txt
@@ -1,23 +1,31 @@
 cmake_minimum_required(VERSION ${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION}.${CMAKE_PATCH_VERSION} )
 
 # Inspired by CMakeDetermineCompilerABI.cmake to set CMAKE_LIBRARY_ARCHITECTURE
+set(LIB_ARCH_LIST)
 if(CMAKE_LIBRARY_ARCHITECTURE_REGEX)
-  if(NOT DEFINED CMAKE_LIBRARY_ARCHITECTURE)
-    file(GLOB implicit_dirs RELATIVE /lib /lib/*-linux-gnu* )
-    foreach(dir ${implicit_dirs})
-      if("${dir}" MATCHES "${CMAKE_LIBRARY_ARCHITECTURE_REGEX}")
-        set(CMAKE_LIBRARY_ARCHITECTURE "${dir}")
-        break()
-      endif()
-    endforeach()
-  endif()
+  file(GLOB implicit_dirs RELATIVE /lib /lib/*-linux-gnu* )
+  foreach(dir ${implicit_dirs})
+    if("${dir}" MATCHES "${CMAKE_LIBRARY_ARCHITECTURE_REGEX}")
+      list(APPEND LIB_ARCH_LIST "${dir}")
+    endif()
+  endforeach()
 endif()
-
-find_package("${NAME}" QUIET)
 
 set(PACKAGE_FOUND FALSE)
 set(_packageName "${NAME}")
 string(TOUPPER "${_packageName}" PACKAGE_NAME)
+
+while(TRUE)
+  find_package("${NAME}" QUIET)
+
+  list(LENGTH LIB_ARCH_LIST LIST_LEN)
+  if(${_packageName}_FOUND OR ${PACKAGE_NAME}_FOUND OR LIST_LEN LESS_EQUAL 0)
+    break()
+  endif()
+
+  list(GET       LIB_ARCH_LIST 0 CMAKE_LIBRARY_ARCHITECTURE)
+  list(REMOVE_AT LIB_ARCH_LIST 0)
+endwhile()
 
 if(${_packageName}_FOUND  OR  ${PACKAGE_NAME}_FOUND)
   set(PACKAGE_FOUND TRUE)

